### ISSUE TITLE
Implement /proc/<pid>/maps

### ISSUE THREAD

**Issue by Francis-Liu** [deislabs/mystikos#250]
This issue emerges as I was testing a .NET Azure SDK [sample](https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/keyvault/samples/getcert) that happens to use `System.CommandLine`, [`InvokeAsync()`](https://github.com/Azure/azure-sdk-for-net/blob/850df4dc21246f6d216e8ed472313b741a8362b7/sdk/keyvault/samples/getcert/Program.cs#L211). I narrowed down the source of the crash to [this](https://github.com/dotnet/command-line-api/blob/722ada8a237d940eb6234792bfa054e1c351129d/src/System.CommandLine/Builder/CommandLineBuilderExtensions.cs#L181) line of code in : 
```
var currentProcessFullPath = Diagnostics.Process.GetCurrentProcess().MainModule.FileName;
```

To reproduce it, I have provided a bare-bones program:
```
using System;
using System.Diagnostics;

namespace processMainModule
{
    class Program
    {
        static void Main(string[] args)
        {
            Process p = Process.GetCurrentProcess();
            var moduleName = p.MainModule.ModuleName;
            Console.WriteLine(moduleName);
        }
    }
}
```
Running the program in Mystikos with either sgx or Linux target will generate the following general error:
```
2021-04-07T00:48:09+0000.652380Z [(H)ERROR] tid(0x7f03e09beb80) | :OE_ENCLAVE_ABORTING [/home/feliu/workspace1/mystikos/third_party/openenclave/openenclave/host/calls.c:_call_enclave_function_impl:56]
/home/feliu/workspace1/dotnet/processMainModule/myst/bin/processMainModule: error: failed to enter enclave: result=OE_ENCLAVE_ABORTING
```
Debugger will show the main thread crushes with "Segmentation fault". The key etrace shows the thread tries to open "/proc/$PID/maps".
```
=== SYS_open(path="/proc/101/maps" flags=02000000 mode=00): tid=101
ERAISE: ramfs.c(626): _path_to_inode_realpath: errno=2: ENOENT
ERAISE: ramfs.c(655): _path_to_inode: errno=2: ENOENT
ERAISE: ramfs.c(825): _fs_open: errno=2: ENOENT
ERAISE: syscall.c(710): myst_syscall_open: errno=2: ENOENT
    SYS_open(): return=-ENOENT(-2): tid=101
```
Some internet search reveals "System.Diagnostics.Process.MainModule relies on the main module being the first module listed in the /proc/self/maps" ([source](https://github.com/dotnet/runtime/issues/23568)).

It is known to me that (currently) Mystikos doesn't support "/proc/$PID/maps" in its filesystem. I open this ticket to track:
1. Root causing why missing "/proc/$PID/maps" caused main thread to receive segmentation fault.
2. Better error handling. Including better error message to help user make sense of failure.

**Comment by bodzhang** [comment]
Adding a [FirstChanceException Notifier](https://docs.microsoft.com/en-us/dotnet/framework/app-domains/how-to-receive-first-chance-exception-notifications) and printout in @Francis-Liu simple app main(), I can see output "FirstChanceException event raised in HelloWorld: Could not find file '/proc/101/maps'." , on `var moduleName = p.MainModule.ModuleName;`

**Comment by jxyang** [comment]
I believe the exception is throw here: https://github.com/dotnet/corefx/blob/f3a4e25a070c2d48592c0ef6a7d48575eb6057df/src/Common/src/Interop/Linux/procfs/Interop.ProcFsStat.cs#L140

Corefx chooses ignore the exception and return an empty list.

FirstChanceException Notifier probably reports it anyway before the code above receives the exception.

**Comment by Francis-Liu** [comment]
- `MainModule` returns `null`. https://github.com/dotnet/corefx/blob/release/3.1/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L236
```
        public ProcessModule MainModule
        {
            get
            {
                ProcessModuleCollection pmc = Modules;
                return pmc.Count > 0 ? pmc[0] : null;
            }
        }
```

- We can verify this by below code sample:
```
using System;
using System.Diagnostics;

namespace processMainModule
{
    class Program
    {
        static void Main(string[] args)
        {
            Process p = Process.GetCurrentProcess();
            var mainModule = p.MainModule;
            Debug.Assert(p.MainModule == null);
            Console.WriteLine("After Debug.Assert MainModule == null");
            var moduleName = mainModule.ModuleName;
            Console.WriteLine(moduleName);
        }
    }
}
```

- There is no further syscall when the managed code access a null.
```
...
=== SYS_write(fd=24 buf=0x7fffc8c0d048 count=38): tid=101
After Debug.Assert MainModule == null
    SYS_write(): return=38(26): tid=101

Thread 1 "myst" received signal SIGSEGV, Segmentation fault.
0x00007fffc77762d9 in ?? ()
```

- Note that if we assign `mainModule=null` on host, we will get the following exception.
```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at processMainModule.Program.Main(String[] args) in /home/feliu/workspace1/dotnet/processMainModule2/Program.cs:line 13
Aborted (core dumped)
```

**Comment by vtikoo** [comment]
This is also blocking running aspnetcore tests.

**Comment by vtikoo** [comment]
Closed via #343
